// hubot/index.js
// Generated by CoffeeScript 1.7.1
(function() {
  var Adapter, Brain, CatchAllMessage, EnterMessage, LeaveMessage, Listener, Response, Robot, TextListener, TextMessage, TopicMessage, User, _ref, _ref1;

  User = require('./src/user');

  Brain = require('./src/brain');

  Robot = require('./src/robot');

  Adapter = require('./src/adapter');

  Response = require('./src/response');

  _ref = require('./src/listener'), Listener = _ref.Listener, TextListener = _ref.TextListener;

  _ref1 = require('./src/message'), TextMessage = _ref1.TextMessage, EnterMessage = _ref1.EnterMessage, LeaveMessage = _ref1.LeaveMessage, TopicMessage = _ref1.TopicMessage, CatchAllMessage = _ref1.CatchAllMessage;

  module.exports = {
    User: User,
    Brain: Brain,
    Robot: Robot,
    Adapter: Adapter,
    Response: Response,
    Listener: Listener,
    TextListener: TextListener,
    TextMessage: TextMessage,
    EnterMessage: EnterMessage,
    LeaveMessage: LeaveMessage,
    TopicMessage: TopicMessage,
    CatchAllMessage: CatchAllMessage
  };

  module.exports.loadBot = function(adapterPath, adapterName, enableHttpd, botName) {
    return new Robot(adapterPath, adapterName, enableHttpd, botName);
  };

}).call(this);
